osu.Server.Spectator/Hubs/ScoreUploader.cs

```diff
@@ -0,0 +1,163 @@
+using osu.Server.Spectator.Storage;
+using Timer = System.Timers.Timer;
+
+namespace osu.Server.Spectator.Hubs
+{
+    public class ScoreUploader : IEntityStore, IDisposable
+    {
+        /// <summary>
+        /// Amount of time (in milliseconds) between checks for pending score uploads.
+        /// </summary>
+        public double UploadInterval
+        {
+            get => timer.Interval;
+            set => timer.Interval = value;
+        }
+
+        /// <summary>
+        /// Amount of time (in milliseconds) before any individual score times out if a score ID hasn't been set.
+        /// This can happen if the user forcefully terminated the game before the API score submission request is sent, but after EndPlaySession() has been invoked.
+        /// </summary>
+        public double TimeoutInterval = 30000;
+
+        private bool shouldSaveReplays => Environment.GetEnvironmentVariable("SAVE_REPLAYS") == "1";
+
+        private readonly ConcurrentQueue<UploadItem> queue = new ConcurrentQueue<UploadItem>();
+        private readonly IDatabaseFactory databaseFactory;
+        private readonly IScoreStorage scoreStorage;
+        private readonly Timer timer;
+        private readonly CancellationTokenSource timerCancellationSource;
+        private readonly CancellationToken timerCancellationToken;
+
+        public ScoreUploader(IDatabaseFactory databaseFactory, IScoreStorage scoreStorage)
+        {
+            this.databaseFactory = databaseFactory;
+            this.scoreStorage = scoreStorage;
+
+            timerCancellationSource = new CancellationTokenSource();
+            timerCancellationToken = timerCancellationSource.Token;
+
+            timer = new Timer(5000);
+            timer.AutoReset = false;
+            timer.Elapsed += (_, _) => Task.Run(Flush);
+            timer.Start();
+        }
+
+        /// <summary>
+        /// Enqueues a new score to be uploaded.
+        /// </summary>
+        /// <param name="token">The score's token.</param>
+        /// <param name="score">The score.</param>
+        public void Enqueue(long token, Score score)
+        {
+            if (!shouldSaveReplays)
+                return;
+
+            Interlocked.Increment(ref remainingUsages);
+
+            var cancellation = new CancellationTokenSource();
+            cancellation.CancelAfter(TimeSpan.FromMilliseconds(TimeoutInterval));
+
+            queue.Enqueue(new UploadItem(token, score, cancellation));
+        }
+
+        /// <summary>
+        /// Flushes all pending uploads.
+        /// </summary>
+        public async Task Flush()
+        {
+            try
+            {
+                timer.Stop();
+
+                if (queue.IsEmpty)
+                    return;
+
+                Console.WriteLine($"Beginning upload of {queue.Count} scores");
+
+                using (var db = databaseFactory.GetInstance())
+                {
+                    int countToTry = queue.Count;
```

Comment:
I'm not sure about using a queue here rather than a `Bucket` or `HashSet`, feels a bit weird to dequeue then requeue potentially hundreds or thousands of items each loop.

Reply:
In practice, this is running on two threads - the signalr thread and the timer thread. The simplest I can come up with this this:

```diff
diff --git a/osu.Server.Spectator/Hubs/ScoreUploader.cs b/osu.Server.Spectator/Hubs/ScoreUploader.cs
index f45f332..1b20417 100644
--- a/osu.Server.Spectator/Hubs/ScoreUploader.cs
+++ b/osu.Server.Spectator/Hubs/ScoreUploader.cs
@@ -3,6 +3,7 @@

 using System;
 using System.Collections.Concurrent;
+using System.Collections.Generic;
 using System.Threading;
 using System.Threading.Tasks;
 using osu.Game.Scoring;
@@ -32,7 +33,7 @@ namespace osu.Server.Spectator.Hubs

         private bool shouldSaveReplays => Environment.GetEnvironmentVariable("SAVE_REPLAYS") == "1";

-        private readonly ConcurrentQueue<UploadItem> queue = new ConcurrentQueue<UploadItem>();
+        private readonly List<UploadItem> queue = new List<UploadItem>();
         private readonly IDatabaseFactory databaseFactory;
         private readonly IScoreStorage scoreStorage;
         private readonly Timer timer;
@@ -68,7 +69,7 @@ namespace osu.Server.Spectator.Hubs
             var cancellation = new CancellationTokenSource();
             cancellation.CancelAfter(TimeSpan.FromMilliseconds(TimeoutInterval));

-            queue.Enqueue(new UploadItem(token, score, cancellation));
+            enqueue(new UploadItem(token, score, cancellation));
         }

         /// <summary>
@@ -80,26 +81,26 @@ namespace osu.Server.Spectator.Hubs
             {
                 timer.Stop();

-                if (queue.IsEmpty)
+                UploadItem[] items;
+
+                lock (queue)
+                    items = queue.ToArray();
+
+                if (items.Length == 0)
                     return;

-                Console.WriteLine($"Beginning upload of {queue.Count} scores");
+                Console.WriteLine($"Beginning upload of {items.Length} scores");

                 using (var db = databaseFactory.GetInstance())
                 {
-                    int countToTry = queue.Count;
-
-                    for (int i = 0; i < countToTry; i++)
+                    foreach (var item in items)
                     {
-                        if (!queue.TryDequeue(out var item))
-                            continue;
-

```

Is that what you're expecting? If not, perhaps it's easier if you implement it yourself.

Reply:
Nah, that's not really any better. I'm just going to turn a blind eye here because it's likely going to change in the future and what you have, while isn't pretty/efficient, does work.